### PR TITLE
Add SlowRide

### PR DESCRIPTION
**Pull request by jgaskins**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR adds support for and some demo code to show the usage of `SlowRide`, a gem that will let us automatically turn off feature flags so we can enable them with more confidence — if the feature-flagged code breaks too much, it will be disabled automatically. This removes the need for a human being to monitor error rates and respond manually.

## QA Instructions, Screenshots, Recordings

The demo code in [`config/initializers/slow_ride.rb`](https://github.com/forem/forem/blob/e2ebd7d4b3e452b7aa0c782301fe985b16db3b6d/config/initializers/slow_ride.rb#L11-L37) contains QA instructions.

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: Just adding initial support and demo code
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

File: config/initializers/slow_ride.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  Redis.new(url: ENV["REDIS_SLOW_RIDE_URL"] || ENV["REDIS_URL"])
+end
+
+# 20% Failure rate
+# 100 checks
+# Reset after it hasn't been checked in 5 minutes
+MY_FEATURE = SlowRide::Redis.new(:my_feature, failure_threshold: 0.2, minimum_checks: 100, max_duration: 5.minutes) do
+  Flipper.disable :my_feature
+  Rails.logger.warn "Flipper my_feature disabled"
+end
```
Comment (jgaskins) [inline config/initializers/slow_ride.rb:14]: The intent is that our SlowRide features are all defined in this file like this one is. They don't have to be `SCREAMING_SNAKE_CASE`, but that's a common convention in Ruby.

File: config/initializers/slow_ride.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#   check_slow_ride { raise "hell" } rescue nil
+#   Flipper.enabled?(:my_feature) # => false, automatically disabled
+def check_slow_ride(&block)
+  if FeatureFlag.enabled?(:my_feature)
+    # Feature flag is on
+    MY_FEATURE.check(&block)
+  else
+    # Do the normal thing, which in this case is nothing
+  end
+end
```
Comment (jgaskins) [inline config/initializers/slow_ride.rb:37]: This code isn't intended to be merged. It's more a talking point than anything.

File: config/initializers/slow_ride.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#   19.times { check_slow_ride { raise "hell" } rescue nil }
+#   Flipper.enabled?(:my_feature) # => true, we haven't reached the failure threshold or the min checks yet
+#
+#   # One more error
+#   check_slow_ride { raise "hell" } rescue nil
+#   Flipper.enabled?(:my_feature) # => false, automatically disabled
+def check_slow_ride(&block)
+  if FeatureFlag.enabled?(:my_feature)
+    # Feature flag is on
+    MY_FEATURE.check(&block)
```
Comment (maestromac) [inline config/initializers/slow_ride.rb:33]: I like it! what's the chance of this detecting false negatives? and what would be the process of restarting slow-ride on the same :feature?  I would imagine a slow-ride experiment is tied to the :name

Comment (jgaskins) [reply config/initializers/slow_ride.rb:33]: @maestromac

> what's the chance of this detecting false negatives?

Right now it only fails a check on exception. I was also considering other failure modes that don't necessarily result in an exception but I wasn't sure how to represent them. Maybe something like this?

```ruby
MY_FEATURE.check do |check|
  result = do_some_work

  if result == :not_what_we_were_hoping_for
    check.fail!
  end

  result
end
```

> what would be the process of restarting slow-ride on the same :feature?

The check deletes itself after invoking the block, so if we follow the convention of disabling the feature flag on failure, restarting the check is simply a matter of re-enabling the feature flag.

If we don't, the block will continue to be invoked every time it meets the requirements. So, for example, the block could be invoked every 1000 requests (`minimum_checks`) as long as the failure rate remains above 20% (`failure_threshold`). I wouldn't want to use this as a common pattern, but I'm sure scenarios exist where this is useful.

>  I would imagine a slow-ride experiment is tied to the :name

Yes, exactly. It's represented in Redis as `slow-ride:#{name}:checked` and `...:failed`.

**Comment by citizen428** [comment]
Hmm, sorry, I completely missed this PR. I'm a bit uncertain how I feel about this feature. I definitely agree on the pain point outlined in the gem's README:

> [A] human being has to pay attention to error rates or, at the very least, respond to alerts about an increase in those error rates.

What makes me a bit uneasy is the automatic disabling though. This could lead to funky PR situations where we announce enabling a feature in a changelog post but by the time people come around it has been automatically disabled. But then, that's probably not worse than the feature being broken.

**Comment by jgaskins** [comment]
@citizen428 Indeed. The purpose is to let the robots do what a human being would do in the same situation because they do it faster. For example, in the scenario you mentioned, we wouldn't keep a broken feature live if it was failing 75% of the time. Manually disabling the feature flag vs automatically disabling it has the same end result.

Disabling the flag also doesn't need to be silent. Being surprised that the feature flag was disabled would likely be pretty bad because then you misinterpret any stats on your feature, including success/failure. Instead, inside the block for the feature check could also be code to send a notification:

```ruby
slack = Slack::Notifier.new(some_config)

MY_FEATURE = SlowRide::Redis.new(:my_feature, failure_threshold: 0.1) do
  FeatureFlag.disable :my_feature
  slack.ping("Feature flag `my_feature` has been automatically disabled due to failures exceeding 10% /cc @my-team")
end
```

**Comment by citizen428** [comment]
Sounds great @jgaskins, thanks for clarifying! A notification in the block seems like the right thing to do, 👍

**Review by citizen428 — APPROVED** [review]

**Comment by msarit** [comment]
@jgaskins Hey hey! Status of this PR?

**Comment by jgaskins** [comment]
There still seems to be a lot more work to do to get alignment on feature flags as a concept, so I probably jumped the gun by going straight to an implementation.

**Pull request closed without merging**